File: instrumentation/wireframe/src/main/java/io/opentelemetry/android/mobile/instrumentation/WireframeInstrumentation.kt
// Copyright 2025 The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

package io.opentelemetry.android.mobile.instrumentation

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.logs.Logger
import io.opentelemetry.api.logs.Severity
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.context.Context
import java.lang.ref.WeakReference

/**
 * Captures a lightweight JSON wireframe of the view hierarchy at trigger points.
 *
 * A wireframe is a compact (~1–5 KB) structural snapshot of every visible [View] in the
 * current activity — class name, bounds, resource ID, clickable state — enough to
 * reconstruct a schematic rendering of the screen without any pixel data.
 *
 * **Trigger modes:**
 * - **Screen transition:** when [WireframeConfig.captureOnScreenView] is true, a wireframe is
 *   captured after each activity/fragment resume (post-layout, so geometry is final).
 * - **Tap:** when [WireframeConfig.captureOnTap] is true, captures after each tap event.
 * - **Error:** when [WireframeConfig.captureOnError] is true, captures on uncaught exceptions.
 * - **Manual:** call [captureWireframe] at any time.
 *
 * **Privacy:** wireframes contain geometry and view types only. Text content is never included
 * unless [WireframeConfig.includeTextHints] is enabled (and even then, only hint/placeholder
 * text — never user-entered values).
 *
 * **Journey replay:** because wireframes are so small, they can be captured on every screen
 * transition and stitched together in a dashboard to visualize the user's path through the app.
 */
@Incubating
class WireframeInstrumentation(
    private val config: WireframeConfig = WireframeConfig()
) : MobileInstrumentation, WindowEventListener {

    override val instrumentationName = "io.opentelemetry.android.mobile.wireframe"

    private var ctx: InstrumentationContext? = null
    private var logger: Logger? = null
    private var tracer: Tracer? = null
    private var uiTelemetryMode: UiTelemetryMode = UiTelemetryMode.EVENTS
    private var application: Application? = null
    private var callbacks: Application.ActivityLifecycleCallbacks? = null
    private var previousExceptionHandler: Thread.UncaughtExceptionHandler? = null

    @Volatile private var currentActivity: WeakReference<Activity>? = null

    private val mainHandler = Handler(Looper.getMainLooper())
    private val rateLimiter = RateLimiter(config.maxCapturesPerMinute)

    // Sequence number for ordering wireframes within a session.
    @Volatile private var sequenceNumber: Long = 0

    // Content-hash dedup state. lastEmittedHash is the SHA-256 of the most-recently
    // EMITTED wireframe JSON (not just the most recently captured). lastEmittedId is
    // the public mobile.wireframe.id we attached to it, so dedup emits a lightweight
    // ui.wireframe.ref pointing at that id instead of the full JSON payload.
    @Volatile private var lastEmittedHash: String? = null
    @Volatile private var lastEmittedId: String? = null

    /** Current wireframe id, or null if no wireframe has been emitted yet in this session. */
    fun currentWireframeId(): String? = lastEmittedId

    override fun install(application: Application, context: InstrumentationContext) {
        if (!config.enabled) return

        this.application = application
        this.ctx = context
        this.logger = context.logger(instrumentationName)
        this.tracer = context.tracer(instrumentationName)
        this.uiTelemetryMode = context.uiTelemetryMode

        val cb = object : Application.ActivityLifecycleCallbacks {
            override fun onActivityResumed(activity: Activity) {
                currentActivity = WeakReference(activity)
                if (config.captureOnScreenView) {
                    // Capture after layout is complete so geometry is final.
                    capturePostLayout(activity, "screen_view")
                }
            }
            override fun onActivityPaused(activity: Activity) {
                if (currentActivity?.get() === activity) {
                    currentActivity = null
                }
            }
            override fun onActivityCreated(a: Activity, b: Bundle?) {}
            override fun onActivityStarted(a: Activity) {}
            override fun onActivityStopped(a: Activity) {}
            override fun onActivitySaveInstanceState(a: Activity, b: Bundle) {}
            override fun onActivityDestroyed(a: Activity) {}
        }
        callbacks = cb
        application.registerActivityLifecycleCallbacks(cb)

        // Listen for tap events if configured.
        if (config.captureOnTap) {
            context.windowEventHub.addListener(this)
        }

        // Auto-capture on uncaught exceptions.
        if (config.captureOnError) {
            previousExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
            Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
                try {
                    captureWireframeSync("error")
                } catch (_: Exception) {
                    // Best-effort.
                }
                previousExceptionHandler?.uncaughtException(thread, throwable)
            }
        }
    }

    override fun uninstall() {
        callbacks?.let { application?.unregisterActivityLifecycleCallbacks(it) }
        callbacks = null

        if (config.captureOnTap) {
            ctx?.windowEventHub?.removeListener(this)
        }

        if (config.captureOnError && previousExceptionHandler != null) {
            Thread.setDefaultUncaughtExceptionHandler(previousExceptionHandler)
            previousExceptionHandler = null
        }

        currentActivity = null
        application = null
        ctx = null
        logger = null
        tracer = null
        uiTelemetryMode = UiTelemetryMode.EVENTS
        rateLimiter.reset()
    }

    override fun onTouchEvent(event: MotionEvent, window: Window) {
        if (!config.captureOnTap) return
        if (event.actionMasked != MotionEvent.ACTION_UP) return

        val activity = currentActivity?.get() ?: return
        val parentContext = Context.current()
        // Small delay to let any resulting layout changes settle.
        mainHandler.postDelayed({ captureFromActivity(activity, "tap", parentContext) }, 100)
    }

    /**
     * Manually capture a wireframe of the current foreground activity.
     *
     * Captures `Context.current()` at call time so the resulting log record
     * carries the active span's trace_id. This stitches the wireframe to
     * the journey span the user was in when it was triggered — see
     * [USER_JOURNEY_CAPTURES_EPIC](docs/epics/USER_JOURNEY_CAPTURES_EPIC.md).
     *
     * @param trigger Describes what triggered the capture (e.g., "manual",
     *   "button_press", "journey_start").
     */
    fun captureWireframe(trigger: String = "manual") {
        if (!config.enabled) return
        if (trigger.startsWith("policy_") && !config.captureOnPolicyMatch) {
            return
        }
        val activity = currentActivity?.get() ?: run {
            Log.d(TAG, "No foreground activity, skipping wireframe capture")
            return
        }
        captureFromActivity(activity, trigger, Context.current())
    }

    /** Synchronous capture — used from crash handler. */
    private fun captureWireframeSync(trigger: String) {
        if (!config.enabled) return
        val activity = currentActivity?.get() ?: return
        if (!rateLimiter.tryAcquire()) return

        val parentContext = Context.current()
        try {
            val rootView = activity.window?.decorView?.rootView ?: return
            val screenName = activity.javaClass.simpleName
            val node = buildTree(rootView, 0)
            emitWireframe(node, screenName, trigger, parentContext)
        } catch (e: Exception) {
            Log.w(TAG, "Sync wireframe capture failed: ${e.message}")
        }
    }

    /**
     * Schedule a capture after the next layout pass, ensuring view geometry is final.
     */
    private fun capturePostLayout(activity: Activity, trigger: String) {
        val rootView = activity.window?.decorView?.rootView ?: return
        val parentContext = Context.current()
        rootView.post {
            captureFromActivity(activity, trigger, parentContext)
        }
    }

    private fun captureFromActivity(activity: Activity, trigger: String, parentContext: Context) {
        if (!rateLimiter.tryAcquire()) {
            Log.d(TAG, "Wireframe rate limit exceeded, skipping capture")
            return
        }

        try {
            val rootView = activity.window?.decorView?.rootView ?: return
            if (rootView.width <= 0 || rootView.height <= 0) return

            val screenName = activity.javaClass.simpleName
            val node = buildTree(rootView, 0)
            emitWireframe(node, screenName, trigger, parentContext)
        } catch (e: Exception) {
            Log.w(TAG, "Wireframe capture failed: ${e.message}")
        }
    }

    /**
     * Recursively build a [WireframeNode] tree from the view hierarchy.
     */
    internal fun buildTree(view: View, depth: Int): WireframeNode {
        val location = IntArray(2)
        view.getLocationInWindow(location)

        val bounds = intArrayOf(
            location[0],
            location[1],
            location[0] + view.width,
            location[1] + view.height
        )

        val type = viewTypeName(view)

        val id = if (config.includeResourceIds) {
            resourceIdName(view)
        } else null

        val hint = if (config.includeTextHints) {
            extractHint(view)
        } else null

        val cd = if (config.includeContentDescription) {
            view.contentDescription?.toString()
        } else null

        val clickable = if (config.includeClickableState) view.isClickable else null
        val enabled = if (config.includeClickableState) view.isEnabled else null

        // Truncate at max depth.
        if (depth >= config.maxDepth) {
            val hasChildren = view is ViewGroup && view.childCount > 0
            return WireframeNode(
                type = type, bounds = bounds, id = id, hint = hint,
                contentDescription = cd, clickable = clickable, enabled = enabled,
                truncated = hasChildren
            )
        }

        val children = if (view is ViewGroup) {
            val list = mutableListOf<WireframeNode>()
            for (i in 0 until view.childCount) {
                val child = view.getChildAt(i)
                if (child.visibility == View.VISIBLE) {
                    list.add(buildTree(child, depth + 1))
                }
            }
            list
        } else emptyList()

        return WireframeNode(
            type = type, bounds = bounds, id = id, hint = hint,
            contentDescription = cd, clickable = clickable, enabled = enabled,
            children = children
        )
    }

    /**
     * Return a short, recognizable type name for the view.
     * Uses simple class name but strips common suffixes for compactness.
     */
    private fun viewTypeName(view: View): String {
        val name = view.javaClass.simpleName
        // For anonymous/generated classes, walk up to the first named superclass.
        if (name.isEmpty() || name.contains('$')) {
            var cls: Class<*> = view.javaClass.superclass ?: return "View"
            while (cls.simpleName.isEmpty() || cls.simpleName.contains('$')) {
                cls = cls.superclass ?: return "View"
            }
            return cls.simpleName
        }
        return name
    }

    /**
     * Extract the Android resource ID name (e.g., "btn_book"), or null.
     */
    private fun resourceIdName(view: View): String? {
        val id = view.id
        if (id == View.NO_ID || id == 0) return null
        return try {
            val name = view.resources.getResourceEntryName(id)
            // Filter out auto-generated IDs from framework.
            if (name.startsWith("android:")) null else name
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Extract hint/placeholder text — never user-entered content.
     */
    private fun extractHint(view: View): String? {
        return when (view) {
            is EditText -> view.hint?.toString()
            is TextView -> {
                // Only include hint, not actual text — for privacy.
                view.hint?.toString()
            }
            else -> null
        }
    }

    private fun emitWireframe(
        node: WireframeNode,
        screenName: String,
        trigger: String,
        parentContext: Context
    ) {
        val context = ctx ?: return
        val sessionProvider = context.sessionProvider

        val json = node.toJson().toString()
        val hash = sha256Hex(json)
        val seq = sequenceNumber++

        // Content-hash dedup: if the wireframe hasn't changed, emit a lightweight ref log
        // pointing at the previously emitted full wireframe instead of resending the JSON.
        // Saves the 1–5 KB payload on no-op screen-resume / tap captures.
        if (config.dedupeByContentHash && hash == lastEmittedHash && lastEmittedId != null) {
            val refAttrs = Attributes.builder()
                .put(MobileSemconv.SESSION_ID, sessionProvider.getSessionId())
                .put(MobileSemconv.VIEW_ID, sessionProvider.getViewId())
                .put(MobileSemconv.SCREEN_NAME, screenName)
                .put(MobileSemconv.WIREFRAME_TRIGGER, trigger)
                .put(MobileSemconv.WIREFRAME_SEQUENCE, seq)
                .put(MobileSemconv.WIREFRAME_ID, lastEmittedId!!)
                .build()
            emitUiTelemetry(MobileSemconv.UI_WIREFRAME_REF, refAttrs, parentContext)
            return
        }

        // First emit or content changed — send the full payload. The id is the hash itself
        // so consumers can correlate ref logs to the originating wireframe deterministically.
        lastEmittedHash = hash
        lastEmittedId = hash

        val attrs = Attributes.builder()
            .put(MobileSemconv.SESSION_ID, sessionProvider.getSessionId())
            .put(MobileSemconv.VIEW_ID, sessionProvider.getViewId())
            .put(MobileSemconv.SCREEN_NAME, screenName)
            .put(MobileSemconv.WIREFRAME_ID, hash)
            .put(MobileSemconv.WIREFRAME_TRIGGER, trigger)
            .put(MobileSemconv.WIREFRAME_SEQUENCE, seq)
            .put(MobileSemconv.WIREFRAME_SIZE_BYTES, json.length.toLong())
            .put(MobileSemconv.WIREFRAME_NODE_COUNT, countNodes(node).toLong())
            .put(MobileSemconv.WIREFRAME_DATA, json)
            .build()

        emitUiTelemetry(MobileSemconv.UI_WIREFRAME, attrs, parentContext)
    }

    private fun sha256Hex(input: String): String {
        val md = java.security.MessageDigest.getInstance("SHA-256")
        val bytes = md.digest(input.toByteArray(Charsets.UTF_8))
        // Hex-encode (lowercase). Short enough that we can pass it as an attribute
        // value cheaply, long enough to make accidental collisions vanishingly rare.
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            sb.append(((b.toInt() shr 4) and 0xF).toString(16))
            sb.append((b.toInt() and 0xF).toString(16))
        }
        return sb.toString()
    }

    private fun emitUiTelemetry(name: String, attrs: Attributes, parentContext: Context) {
        when (uiTelemetryMode) {
            UiTelemetryMode.EVENTS -> logger?.logRecordBuilder()
                ?.setContext(parentContext)
                ?.setBody(name)?.setSeverity(Severity.INFO)?.setAllAttributes(attrs)?.emit()
            UiTelemetryMode.SPANS -> tracer
                ?.spanBuilder(name)?.setParent(parentContext)
                ?.setSpanKind(SpanKind.INTERNAL)?.startSpan()
                ?.apply { setAllAttributes(attrs); end() }
            UiTelemetryMode.BOTH -> {
                logger?.logRecordBuilder()
                    ?.setContext(parentContext)
                    ?.setBody(name)?.setSeverity(Severity.INFO)?.setAllAttributes(attrs)?.emit()
                tracer
                    ?.spanBuilder(name)?.setParent(parentContext)
                    ?.setSpanKind(SpanKind.INTERNAL)?.startSpan()
                    ?.apply { setAllAttributes(attrs); end() }
            }
        }
    }

    /**
     * Test seam: emit a synthetic wireframe log without going through the
     * view-tree pipeline. Validates trace_id propagation through the emit
     * path without needing a real Activity.
     */
    internal fun emitForTesting(
        trigger: String,
        screenName: String,
        parentContext: Context = Context.current()
    ) {
        val context = ctx ?: return
        val sessionProvider = context.sessionProvider
        val attrs = Attributes.builder()
            .put(MobileSemconv.SESSION_ID, sessionProvider.getSessionId())
            .put(MobileSemconv.VIEW_ID, sessionProvider.getViewId())
            .put(MobileSemconv.SCREEN_NAME, screenName)
            .put(MobileSemconv.WIREFRAME_TRIGGER, trigger)
            .put(MobileSemconv.WIREFRAME_SEQUENCE, sequenceNumber++)
            .build()
        emitUiTelemetry(MobileSemconv.UI_WIREFRAME, attrs, parentContext)
    }

    /** Count total nodes in the tree (for the metadata attribute). */
    private fun countNodes(node: WireframeNode): Int {
        return 1 + node.children.sumOf { countNodes(it) }
    }

    /** Visible for testing. */
    internal val trackedActivity: Activity? get() = currentActivity?.get()
    internal val isInstalled: Boolean get() = callbacks != null

    companion object {
        private const val TAG = "OTel-Wireframe"
    }
}

File: otel-ios-mobile/Sources/WireframeInstrumentation/WireframeInstrumentation.swift
/*
 * Copyright 2025 Barry Solomon
 * SPDX-License-Identifier: Apache-2.0
 */
import CryptoKit
import Foundation
import OpenTelemetryApi
import OTelMobileCore
#if canImport(UIKit)
import UIKit
#endif

public final class WireframeInstrumentation: @unchecked Sendable, TouchEventListener {
    public static let shared = WireframeInstrumentation()

    public let config: WireframeConfig

    private let lock = NSLock()
    private var installed = false
    private var logger: Logger?
    private var tracer: Tracer?
    private var sessionProvider: SessionProvider?
    private var eventHub: TouchEventHub?
    private var rateLimiter: RateLimiter
    private var sequenceNumber: Int64 = 0

    // Content-hash dedup state. `lastEmittedHash` is the SHA-256 of the most
    // recently EMITTED full wireframe JSON (not just the most recently
    // captured). `lastEmittedId` is the public `mobile.wireframe.id` we
    // attached to it, so dedup emits a `ui.wireframe.ref` pointing at that
    // id instead of the full payload. Both guarded by `lock`. Mirrors
    // Android's WireframeInstrumentation dedup state.
    private var lastEmittedHash: String?
    private var lastEmittedId: String?

    /// Current wireframe id, or `nil` if no wireframe has been emitted yet
    /// in this session. Companion to Android's
    /// `WireframeInstrumentation.currentWireframeId()`.
    public func currentWireframeId() -> String? {
        lock.lock(); defer { lock.unlock() }
        return lastEmittedId
    }

    #if canImport(UIKit) && (os(iOS) || os(tvOS))
    private var observers: [NSObjectProtocol] = []
    #endif

    public init(config: WireframeConfig = WireframeConfig()) {
        self.config = config
        self.rateLimiter = RateLimiter(maxPerWindow: config.maxCapturesPerMinute)
    }

    public func install(context: InstrumentationContext) {
        lock.lock()
        if installed || !config.enabled {
            lock.unlock()
            return
        }
        installed = true
        self.logger = context.logger
        self.tracer = context.tracer
        self.sessionProvider = context.sessionProvider
        self.eventHub = context.eventHub
        lock.unlock()

        if config.captureOnTap {
            context.eventHub.addListener(id: "wireframe", listener: self)
        }

        #if canImport(UIKit) && (os(iOS) || os(tvOS))
        if config.captureOnScreenView {
            let nc = NotificationCenter.default
            lock.lock()
            observers.append(nc.addObserver(
                forName: UIScene.didActivateNotification,
                object: nil, queue: .main
            ) { [weak self] _ in
                self?.capture(trigger: "screen_view")
            })
            lock.unlock()
        }
        #endif
    }

    public func uninstall() {
        lock.lock(); defer { lock.unlock() }
        installed = false
        #if canImport(UIKit) && (os(iOS) || os(tvOS))
        let nc = NotificationCenter.default
        for o in observers { nc.removeObserver(o) }
        observers.removeAll()
        #endif
        if config.captureOnTap {
            eventHub?.removeListener(id: "wireframe")
        }
        logger = nil
        tracer = nil
        sessionProvider = nil
        eventHub = nil
        rateLimiter.reset()
    }

    // MARK: - TouchEventListener

    public func onTouchEvent(_ event: TouchEventHub.Event) {
        guard config.captureOnTap, event.type == .touchUp else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) { [weak self] in
            self?.capture(trigger: "tap")
        }
    }

    // MARK: - Capture

    public func capture(trigger: String = "manual") {
        guard config.enabled else { return }
        // Trigger-specific gates: policy-match captures default-on but
        // configurable via WireframeConfig.captureOnPolicyMatch.
        if trigger.hasPrefix("policy_") && !config.captureOnPolicyMatch {
            return
        }
        guard rateLimiter.tryAcquire() else { return }

        #if canImport(UIKit) && (os(iOS) || os(tvOS))
        // captureFromKeyWindow walks UIApplication/UIWindow/UIView, which are
        // main-thread-only. The tap / screen_view triggers already arrive on
        // main, but the error / policy-match / public-API paths can call
        // capture() from a background Task.detached (see
        // MobileLogRecordProcessor). Hop to main for those to avoid off-main
        // UIKit access (undefined behavior / host crash).
        if Thread.isMainThread {
            captureFromKeyWindow(trigger: trigger)
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.captureFromKeyWindow(trigger: trigger)
            }
        }
        #endif
    }

    /// Evaluate the consent gate synchronously on the main thread (the capture
    /// path is already on the main thread). Returns `true` when no gate is
    /// configured or the gate authorizes the capture.
    internal func consentAllows(trigger: String, screenName: String?) -> Bool {
        guard let gate = config.shouldCapture else { return true }
        let context = CaptureContext(
            trigger: CaptureTrigger(rawTrigger: trigger),
            kind: .wireframe,
            screenName: screenName
        )
        return gate(context)
    }

    #if canImport(UIKit) && (os(iOS) || os(tvOS))
    private func captureFromKeyWindow(trigger: String) {
        guard let window = Self.findKeyWindow() else { return }

        // Consent gate: consulted synchronously on the main thread immediately
        // before the view-tree walk. If it denies, skip the capture entirely.
        let screenName = Self.topViewControllerName(in: window)
        guard consentAllows(trigger: trigger, screenName: screenName) else { return }

        let node = buildTree(view: window, depth: 0)
        let json = node.toJson()
        emitWireframe(json: json, nodeCount: node.nodeCount(), screenName: screenName, trigger: trigger, sizeBytes: json.count)
    }

    internal func buildTree(view: UIView, depth: Int) -> WireframeNode {
        let frame = view.frame
        let bounds = [Int(frame.origin.x), Int(frame.origin.y),
                      Int(frame.origin.x + frame.width), Int(frame.origin.y + frame.height)]

        let type = Self.viewTypeName(view)

        // Deterministic redaction: a node is redacted when the shared policy
        // says so (secure UIKit field or explicitly-tagged view, incl. SwiftUI
        // `.dash0Redacted()`). Redacted nodes drop ALL text-bearing fields (the
        // WireframeNode initializer enforces this) and we do NOT descend into
        // their subtree, so no sensitive child text can leak either.
        let isRedacted = Dash0RedactionPolicy.shouldRedact(view, redactAllText: false)

        let interactive: Bool? = config.includeInteractionState ? view.isUserInteractionEnabled : nil
        let enabled: Bool? = config.includeInteractionState ? (view.alpha > 0 && !view.isHidden) : nil

        if isRedacted {
            return WireframeNode(
                type: type, bounds: bounds,
                isInteractive: interactive, isEnabled: enabled,
                redacted: true
            )
        }

        let aid: String? = config.includeAccessibilityIdentifiers ? view.accessibilityIdentifier : nil
        let hint: String? = config.includeTextHints ? extractHint(from: view) : nil
        let label: String? = config.includeContentDescription ? view.accessibilityLabel : nil

        if depth >= config.maxDepth {
            return WireframeNode(
                type: type, bounds: bounds,
                accessibilityIdentifier: aid, hint: hint, accessibilityLabel: label,
                isInteractive: interactive, isEnabled: enabled,
                truncated: !view.subviews.isEmpty
            )
        }

        let children = view.subviews
            .filter { !$0.isHidden }
            .map { buildTree(view: $0, depth: depth + 1) }

        return WireframeNode(
            type: type, bounds: bounds,
            accessibilityIdentifier: aid, hint: hint, accessibilityLabel: label,
            isInteractive: interactive, isEnabled: enabled,
            children: children
        )
    }

    private static func viewTypeName(_ view: UIView) -> String {
        String(describing: Swift.type(of: view))
            .components(separatedBy: ".")
            .last ?? "UIView"
    }

    private func extractHint(from view: UIView) -> String? {
        if let tf = view as? UITextField { return tf.placeholder }
        if let tv = view as? UITextView { return tv.text.isEmpty ? nil : nil }
        return nil
    }

    private static func topViewControllerName(in window: UIWindow) -> String {
        var vc = window.rootViewController
        while let presented = vc?.presentedViewController {
            vc = presented
        }
        if let nav = vc as? UINavigationController {
            vc = nav.topViewController
        }
        if let tab = vc as? UITabBarController {
            vc = tab.selectedViewController
        }
        return vc.map { String(describing: Swift.type(of: $0)) } ?? "Unknown"
    }

    internal static func findKeyWindow() -> UIWindow? {
        if #available(iOS 15.0, *) {
            return UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .flatMap { $0.windows }
                .first { $0.isKeyWindow }
        }
        return nil
    }
    #endif

    private func emitWireframe(json: String, nodeCount: Int, screenName: String, trigger: String, sizeBytes: Int) {
        let hash = Self.sha256Hex(json)

        // Snapshot mutable state under the lock. If dedup is on and the hash
        // matches the previously emitted wireframe, emit a `ui.wireframe.ref`
        // log carrying the prior id only — no JSON payload.
        lock.lock()
        let logger = self.logger
        let session = self.sessionProvider
        let seq = sequenceNumber
        sequenceNumber += 1
        let dedupHit = config.dedupeByContentHash
            && hash == lastEmittedHash
            && lastEmittedId != nil
        let priorId = lastEmittedId
        if !dedupHit {
            lastEmittedHash = hash
            lastEmittedId = hash
        }
        lock.unlock()

        guard let logger = logger else { return }

        if dedupHit, let priorId = priorId {
            var attrs: [String: AttributeValue] = [
                "mobile.wireframe.trigger": .string(trigger),
                "mobile.wireframe.sequence": .int(Int(seq)),
                "mobile.wireframe.id": .string(priorId),
                "screen.name": .string(screenName),
            ]
            if let sid = session?.sessionId {
                attrs["mobile.session.id"] = .string(sid)
            }
            logger.logRecordBuilder()
                .setBody(AttributeValue.string("ui.wireframe.ref"))
                .setSeverity(.info)
                .setAttributes(attrs)
                .emit()
            return
        }

        // First emit or content changed — send the full payload. The id is
        // the hash itself so consumers can correlate ref logs to the
        // originating wireframe deterministically.
        var attrs: [String: AttributeValue] = [
            "mobile.wireframe.trigger": .string(trigger),
            "mobile.wireframe.sequence": .int(Int(seq)),
            "mobile.wireframe.id": .string(hash),
            "mobile.wireframe.size_bytes": .int(sizeBytes),
            "mobile.wireframe.node_count": .int(nodeCount),
            "mobile.wireframe.data": .string(json),
            "screen.name": .string(screenName),
        ]
        if let sid = session?.sessionId {
            attrs["mobile.session.id"] = .string(sid)
        }

        logger.logRecordBuilder()
            .setBody(AttributeValue.string("ui.wireframe"))
            .setSeverity(.info)
            .setAttributes(attrs)
            .emit()
    }

    private static func sha256Hex(_ input: String) -> String {
        let data = Data(input.utf8)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Test seam

    internal func emitForTesting(trigger: String, screenName: String) {
        lock.lock()
        let logger = self.logger
        let session = self.sessionProvider
        let seq = sequenceNumber
        sequenceNumber += 1
        lock.unlock()

        guard let logger = logger else { return }

        var attrs: [String: AttributeValue] = [
            "mobile.wireframe.trigger": .string(trigger),
            "screen.name": .string(screenName),
            "mobile.wireframe.sequence": .int(Int(seq)),
        ]
        if let sid = session?.sessionId {
            attrs["mobile.session.id"] = .string(sid)
        }

        logger.logRecordBuilder()
            .setBody(AttributeValue.string("ui.wireframe"))
            .setSeverity(.info)
            .setAttributes(attrs)
            .emit()
    }

    internal var isInstalled: Bool {
        lock.lock(); defer { lock.unlock() }
        return installed
    }
}
